diesel/src/pg/metadata_lookup.rs

```diff
@@ -24,26 +25,87 @@ impl PgMetadataLookup {
     /// This function should only be used for user defined types, or types which
     /// come from an extension. This function may perform a SQL query to look
     /// up the type. For built-in types, a static OID should be preferred.
-    pub fn lookup_type(&self, type_name: &str) -> PgTypeMetadata {
+    pub fn lookup_type(&self, type_name: &str, schema: Option<&str>) -> PgTypeMetadata {
         let metadata_cache = self.conn.get_metadata_cache();
-        metadata_cache.lookup_type(type_name).unwrap_or_else(|| {
-            use self::pg_type::dsl::*;
-
-            let type_metadata = pg_type
-                .select((oid, typarray))
-                .filter(typname.eq(type_name))
-                .first(&self.conn)
-                .unwrap_or_default();
-            metadata_cache.store_type(type_name, type_metadata);
-            type_metadata
+        let cache_key = PgMetadataCacheKey {
+            schema: schema.map(Cow::Borrowed),
+            type_name: Cow::Borrowed(type_name),
+        };
+
+        metadata_cache.lookup_type(&cache_key).unwrap_or_else(|| {
+            let r = lookup_type(&cache_key, &self.conn);
+
+            match r {
+                Ok(type_metadata) => {
+                    metadata_cache.store_type(cache_key, type_metadata);
+                    type_metadata
+                }
+                Err(_e) => {
```

Comment:
That's a good discussion point. There is no 1.4.x with metadata caching, therefore the current default behaviour is to just run the query every time we need the type oid. Another problem with the old implementation is that it does just silently accept that a type is non-existent and just returned 0 in those cases. Now this works most of the time but lead to surprising bugs in the past (Something like this works everywhere, but for this one query that requires something to be bound as array element).
The question is how to correctly handle cases where the oid does not exist (yet?):
* Just `panic!` , which I believe is a bad solution
* Change `HasSqlType::metadata` to return a result and return an error in such cases? That's a quite large change, not sure if that's worth it. (+ other database systems cannot fail there)
* Log and error and do that what should work most of the time. That's this solution. Now I've gone for the variant that requires that types anytime they are requested, as the type oid could appear, because the type is created dynamically (thought that's quite unlikely).

In general postgres system of allowing to change/add a new type at runtime is quite hard to come by with caching. There will likely always be a case that slip through and breaks things. The only bullet proof variant seems to be to fetch everything each request, but that's obviously bad for performance.

Reply:
> The only bullet proof variant seems to be to fetch everything each request, but that's obviously bad for performance.

[I initially implemented this caching](https://github.com/diesel-rs/diesel/pull/2088) with your help precisely because the performance impact turned out to be extremely high ^^

> I've gone for the variant that requires that types anytime they are requested, as the type oid could appear, because the type is created dynamically (thought that's quite unlikely).

The not-so-unlikely scenario that would be an issue for us is:
 - OID is not found because it's badly declared (e.g. it's in the wrong schema, or it has the wrong name due to a typo)
 - Postgres silent-recovers from the OID not being found and infers it all the time
 - The query costs every time in case it would be found, sliently (or almost, possibly generating lots of logs) having a huge perf impact

So I believe we have to make a choice between:
1. Letting PG attempt to recover without giving it the type OID
2. Re-sending query if OID was not found last time (in case it's created dynamically *after a previous fail*)

Indeed, not doing 1 would prevent the silent recover (forcing a fix -> no silent perf impact), and not doing 2 would prevent re-sending the query (-> no perf impact).

Now as I don't like silent fails I'd favor dropping 1 but it looks like dropping 1 would, as you mentioned, require either of:

> * Just `panic!` , which I believe is a bad solution
> * Change `HasSqlType::metadata` to return a result and return an error in such cases? That's a quite large change, not sure if that's worth it. (+ other database systems cannot fail there)

I don't like panicking in this scenario either, so if changing `HasSqlType::metadata` to return a result is too complex then it looks like we should drop 2 instead.

Reply:
I've gone forward to make this actually return an error if such a unknown type is used somewhere, by wrapping this information as part of the `PgTypeMetadata` struct. This somehow seems to be the correct solution for this problem, even if that's the larger change. As part of this change I've made that struct opaque as this could potentially help with #2087 later.

Another thing we could do here is to use case insensitive comparisons for the type and schema names as postgres does not care about this internally. I think all type names are lower cased internally or something like that.

Reply:
So I haven't had time to look at every usage of TypeMetadata yet so this may be imprecise, but with the error possibly wrapped in `PgTypeMetadata` and only triggering when attempting to read the inner values, I'm not sure I see what guarantees there isn't a scenario where we build `PgTypeMetadata`, possibly resulting in wrapping an error inside (not storing cache), then actually not use it -> not triggering the error. If there was such a scenario, then my scenario at https://github.com/diesel-rs/diesel/pull/2628/files#r564001970 could still happen.

What guarantees that actually attempt to extract an OID out of `PgTypeMetadata` everytime we've built ?
If that is hard to guarantee, maybe we want to make `lookup_type` directly return a `Result`, and `?` it right away everytime ?

Reply:
True, it's possible to just call `SomeSqlType::metadata(lookup)` somewhere without using the result. For diesel itself I would consider that as a bug (and I'm really sure we are using it everywhere), for third party code that's up to whatever they want to do there.
I've added at `#[must_use]` to `PgTypeMetadata` to make it at least warn if we would construct an instance of that type without using it. 

I've opted against making `HasSqlType::metadata()` return as this is a much larger change and all of the other backends have a closed not user extensible set of type metadata.

Reply:
> I've added at `#[must_use]` to `PgTypeMetadata`
> I'm really sure we are using it everywhere

Alright :)